### PR TITLE
VM: Deduplicate the virtiofsd start and stop logic and simplify disk apparmor rules

### PR DESCRIPTION
**Pull request by tomponline**
Removes duplicate virtiofsd start and stop logic to centralise the knowledge for when virtiofsd should be used, rather than duplicating it in both the qemu driver and the disk device packages.

- Also centralises the variable generation for the config drive's virtiofsd socket and pid paths.
- Also explicitly allows the qemu process r/w access to the instance's devices directory. This is to allow access to the virtiofsd socket, and in the future bind-mounted disk directories themselves. I've allowed access to the whole directory rather than explicitly adding boot-time rules individually to allow us to support hot plugged disks in the future.

### CODE REVIEW

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1765]: @stgraber @monstermunchkin does this script exist elsewhere (distrobuilder, ubuntu?) that I should also update?

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1765]: @monstermunchkin thanks, once this is approved i'll go and fix it there too.

**Review by monstermunchkin — APPROVED** [review]

**Comment by stgraber** [comment]
@tomponline we can't do this, there are distros that only support virtiofs and not 9p (centos and derivatives).

**Comment by tomponline** [comment]
> @tomponline we can't do this, there are distros that only support virtiofs and not 9p (centos and derivatives).

Hrm thats a bit of a problem, as using virtiofs for the config drive is a security risk.

**Comment by stgraber** [comment]
@tomponline what's the security risk? I mean, I'd prefer it be read-only but it's size restricted and hidden away on the host.

If we need to fix this, the right way would be to implement what I described yesterday, setup read-only bind-mounts on the host and then export that through virtiofsd.

**Comment by tomponline** [comment]
> @tomponline what's the security risk? I mean, I'd prefer it be read-only but it's size restricted and hidden away on the host.
> 
> If we need to fix this, the right way would be to implement what I described yesterday, setup read-only bind-mounts on the host and then export that through virtiofsd.

Indeed, I'll go an evaluate it (I'd assumed that because the 9p is readonly, the virtio-fs equivalent should be readonly too, otherwise the behaviour is confusing and could (did for me) lead to confusion later).

Also, from the discussion here https://discuss.linuxcontainers.org/t/failure-with-readonly-mount-on-vm/11179/16?u=tomp it seems that the bind mount approach doesn't work (I have also done a cursory reproducer for that but will dig it into more detail to see what is going on).

Also, as part of the testing in https://github.com/lxc/lxc-ci/pull/291#issuecomment-848949045 I've noticed that the bootindex order for devices that don't have an explicit boot.priority (or not root) set are random, so have just added https://github.com/lxc/lxd/pull/8813/commits/345b86a82ad6116860f26c201e9d3ef49186e017 to address. As this was causing the check for the readonly config drive to intermittently fail as the drives were appearing under different /dev/sd{n} names based on the scsi-id (which is being derived from the bootindex).

**Comment by tomponline** [comment]
@stgraber does this mean that https://github.com/lxc/lxd/pull/8810 also needs to be re-worked, as that disables virtio-fs and uses 9p when setting `readonly=true` in the disk device?

**Comment by stgraber** [comment]
Yeah, we'll want to fix that too. It's slightly less problematic since you at least get to boot the instance, just not get your disks mounted but we should apply the same solution to both cases.

**Comment by tomponline** [comment]
@stgraber I'll also add some comments about requiring virtio-fs support for some distros that don't support 9p, as the comments at the moment very much suggest that 9p is the universal fallback.

**Comment by tomponline** [comment]
> Yeah, we'll want to fix that too. It's slightly less problematic since you at least get to boot the instance, just not get your disks mounted but we should apply the same solution to both cases.

OK I think the bind-mount approach is going to be the way to do as we discussed, and then the 9p readonly mode is just an extra layer of security. This way we can then get consistent behaviour across both share types.

**Comment by tomponline** [comment]
@stgraber @monstermunchkin as part of the process to get to fully supporting readonly config drive and VM disks directory shares, I've reworked how we start/stop the virtiofsd process so that the knowledge about when we do and do not use virtiofsd is centralised.

**Comment by tomponline** [comment]
Please hold off on review ATM, there's an issue with apparmor changes in my testing...

**Comment by tomponline** [comment]
@stgraber ready for review

**Pull request merged**
